### PR TITLE
DietPi-Software | Pi-hole: Fix blocking page + add Apache/Nginx configs

### PR DESCRIPTION
**Pull request by MichaIng**
**Status**: Ready
- [x] Add Pi-hole webserver config files to `.conf/dps_93/`
- [x] DietPi-Software | Add/comment 404 handler in Pi-hole webserver configs based on blocking page decision. A separate mod/config for this is overkill, public access blocking needs to be done anyway.
- [x] DietPi-Patch | Remove obsolete index.php => blocking page symlink
- [x] Changelog

**Testing**:
- [x] Upgrade to v6.26 with Pi-hole + Lighttpd
- [x] Upgrade to v6.26 with Pi-hole + Nginx
- [x] Upgrade to v6.26 with Pi-hole + Apache
- [x] Fresh Pi-hole + Lighttpd install on v6.26
- [x] Fresh Pi-hole + Nginx install on v6.26
- [x] Fresh Pi-hole + Apache install on v6.26

**Reference**: https://github.com/MichaIng/DietPi/pull/3054

**Commit list/description**:
+ DietPi-Software | Pi-hole: Fix blocking page implementation which needs to be done via 404 handler. Otherwise only requests without path will show the blocking page. Add configs for each webserver to enable blocking page 404 handler and access permissions.
+ DietPi-Software | Pi-hole: Disable obsolete dhcpcd service. It is installed by Pi-hole installer, to disable DHCP, but on DietPi we use regular way via /etc/network/interfaces by using the "static" option.
+ DietPi-Software | Minor coding and wording

### CODE REVIEW

**Comment by MichaIng** [comment]
@AtkLordOverAll
Finally done. Would you mind to review/test it?

_________________

This was much more work then expected. Every webserver has it's own quirks and behaves different/has different limitations.

Especially Nginx made much trouble:
- It never parses multiple location directives on one request, but only the first or best match. It is still not clear to me in which case it tries/uses the next matching location and in which case not. Try and error... This also means that PHP handling, headers etc must be defined again in every single location directive.
- Because of this the optional admin panel block is a bid ugly... No way to add it as Independent config, it must be included inside the existing admin panel location directive.
- It has it's own regex syntax that differs from what one is used in UNIX/shell.

It is still not perfect. The directory indexes/listing and/or index page function of Nginx and Apache bring some limitations:
- On Apache, by default directory indexes are active, which means that the content of the directory is shown in browser, if no index.php/html page is present. So if some blocked URL has no path, the webserver does not answer with 404 (not found, => blocking page), but the webroot directory is shown instead. If one disables the directory indexes option, then the webserver answers with 403 (forbidden) instead of 404 (not found), so again no blocking page is shown... The only way would be to disable the autoindex mod. But might be used by users for other sub dirs AND most importantly Indexes option that relies on this mod is present in the default Apache configuration files, which then would lead to startup failure...
- On Nginx it is somewhat the other way round. Directory listing is disabled by default. When no index file is present, one as well get a 403 (forbidden) instead of a 404.
- In both cases it would be the perfect solution, if disabled directory indexes/listing + no index file could be answered with 404 => blocking page instead of 403... But there does not seem to be a setting to switch this, so not possible without some complicated rewrite rule overkill...

**Comment by MichaIng** [comment]
Okay I merge this now. Testing can then be done via dev branch. E.g. to create a backup and migrate to dev branch:
```
dietpi-backup 1
G_DEV_BRANCH dev
```

**Comment by alpha-tango-kilo** [comment]
Okay looks good, sorry that ended up being so much more effort for you that expected! As for testing, unfortunately I'm not fluent in any of the three webservers to be able to try and test this for you. I do have Pi-hole installed on one of my devices so if you wanted to give me some instructions for some simple tests I'd be happy to. The reason I initially took up this issue was because I thought it would be easy enough for me to help you with 😅

**Comment by MichaIng** [comment]
@AtkLordOverAll 
One way to test would be to switch to dev branch with the two commands I posted above (first just to have a backup to revert).

To only test the new Lighttpd config (this is your webserver, right?) with blocking page support and access permissions:
```
[[ -L '/var/www/index.php' && $(readlink -f '/var/www/index.php') == '/var/www/html/pihole/index.php' ]] && rm /var/www/index.php
G_RUN_CMD wget "https://raw.githubusercontent.com/MichaIng/DietPi/dev/.conf/dps_93/lighttpd.block_public_admin.conf" -O /etc/lighttpd/conf-available/99-dietpi-pihole-block_public_admin.conf
G_RUN_CMD wget "https://raw.githubusercontent.com/MichaIng/DietPi/dev/.conf/dps_93/lighttpd.pihole.conf" -O /etc/lighttpd/conf-available/99-dietpi-pihole.conf
lighttpd-enable-mod dietpi-pihole
lighttpd-enable-mod dietpi-pihole-block_public_admin
if grep -q 'BLOCKINGMODE=IP-NODATA-AAAA' /etc/pihole/pihole-FTL.conf; then
[[ -f '/var/www/index.lighttpd.html' ]] && mv /var/www/index.lighttpd.html /var/www/index.lighttpd.html.bak
G_CONFIG_INJECT 'server.error-handler-404' 'server.error-handler-404 = "/html/pihole/index.php"' /etc/lighttpd/conf-available/99-dietpi-pihole.conf
fi
```
- This only enables the blocking page, if this was selected already on the initial Pi-hole install.
- It will enable the config to block public access to admin panel.
- So with this, public access to blocking page (your.domain.org/pihole + your.domain.org/html/pihole) and admin panel (your.domain.org/admin + your.domain.org/html/admin) should both report error 403 (permission denied).
- Additionally access to dot files/dirs should be blocked, as well from LAN access:, e.g. your.domain.org/admin/.gitignore + your.domain.org/html/admin/.gitignore

> The reason I initially took up this issue was because I thought it would be easy enough for me to help you with 😅

And I highly appreciate. Your addition serves well, including dialog about blocking public admin panel access etc.
Fix for blocking page implementation was unrelated. Good that we worked on this topic, otherwise I would not have recognised this bug too soon.
And as of the other webservers, it was good to freshen knowledge about their configs and syntax.

So finally all this has been very productive, nothing to feel sorry for 😃.

**Comment by root9191** [comment]
Hello, i have the newest Version 7.1.2 installed and i got Pihole on Apache Server installed. Everything works fine, but if i access a blocked page, nothing happens. I dont get redirected to the blockin page and i dont know why.
I changed the Port to 82, because i have a Reverse Proxy on Port 80. But that could not be the problem?
Maybe you can help me here.

Best Regards

**Comment by MichaIng** [comment]
TL;DR: Do not use it (anymore)!

The blocking page is disabled by default and it is about to be removed from Pi-hole: https://github.com/pi-hole/pi-hole/pull/3910
It would require you to change the [blocking mode](https://docs.pi-hole.net/ftldns/blockingmode/) and use the blocking page as 404 handler, but it did never and can never work on any HTTPS-served ad, so looses more and more purpose in a web where even ads are served more and more via HTTPS. IMO that does not weight out anymore the overhead it implies when every ad implies a full request to your Pi-hole server instead of an immediate give-up by the browser/client to load anything.
